README.rst
=============================
logjuggler
=============================

.. image:: https://travis-ci.org/qba73/logjuggler.png?branch=master
        :target: https://travis-ci.org/qba73/logjuggler


A simple log parser.


Features
--------

* Parsing logs
* Searching logs by: level, session_id, business_id, request_id and date range
* Profiling func executions (calls, time: avg, max, min)

TODO:
--------

* Logging
* Python3 support (Travis)
* Support for multiple filters


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2014-01-14)
++++++++++++++++++

* First release.


test/test_logjuggler.py
"""

Tests for `logjuggler` module.

"""

import datetime
import pytest
from logjuggler import logjuggler


@pytest.fixture
def log_line():
    return "2012-09-13 16:04:22 DEBUG SID:34523 BID:1329 RID:65d33 'Starting new session'"


@pytest.fixture
def log_lines():
    return [
        logjuggler.Log(datetime.datetime(2012, 9, 13, 16, 4, 22),
            level='DEBUG', session_id='34523', business_id='1329',
            request_id='65d33', message='Starting new session'
        ),
        logjuggler.Log(datetime.datetime(2012, 9, 13, 16, 4, 30),
            level='DEBUG', session_id='34523', business_id='1329',
            request_id='54f22', message='Authenticating User'
        ),
        logjuggler.Log(datetime.datetime(2012, 9, 13, 16, 4, 33),
            level='ERROR', session_id='34523', business_id='1345',
            request_id='54ff3', message='Missing Authentication token'
        ),
        logjuggler.Log(datetime.datetime(2012, 9, 13, 16, 4, 22),
            level='DEBUG', session_id='42111', business_id='319',
            request_id='7a323', message='Deleting asset with ID 543234'
        ),
        logjuggler.Log(datetime.datetime(2012, 9, 14, 16, 4, 22),
            level='WARN', session_id='42111', business_id='319',
            request_id='7a323', message='Invalid asset ID'
        ),
    ]


class TestLogLineParsers(object):
    def test_extract_log_message(self, log_line):
        assert logjuggler.log_message(log_line) == 'Starting new session'

    def test_extract_log_level(self, log_line):
        assert logjuggler.log_level(log_line) == 'DEBUG'

    def test_extract_business_id(self, log_line):
        assert logjuggler.business_id(log_line) == '1329'

    def test_extract_request_id(self, log_line):
        assert logjuggler.request_id(log_line) == '65d33'

    def test_extract_session_id(self, log_line):
        assert logjuggler.session_id(log_line) == '34523'

    def test_timestamp_should_be_datetime_object(self, log_line):
        assert isinstance(logjuggler.log_time(log_line), datetime.datetime)

    def test_time_to_iso_should_return_timestamp_in_iso_format(self, log_line):
        assert logjuggler.time_to_iso(logjuggler.log_time(log_line)) ==\
                                                        '2012-09-13 16:04:22'


class TestDateRangeFilter(object):
    def test_log_dates_outside_range(self, log_lines):
        start_test_date = datetime.datetime(2011, 1, 12, 12, 1, 1)
        end_test_date = datetime.datetime(2011, 1, 12, 17, 1, 1)
        test_filter = logjuggler.date_range_filter(
            start_test_date, end_test_date)
        assert list(logjuggler.search_results(test_filter, log_lines)) == []

    def test_line_dates_inside_range(self, log_lines):
        start_test_date = datetime.datetime(2012, 9, 13, 16, 4, 22)
        end_test_date = datetime.datetime(2012, 9, 14, 16, 5, 32)
        test_filter = logjuggler.date_range_filter(
            start_test_date, end_test_date)
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 5

    def test_log_dates_inside_and_outside_range(self, log_lines):
        start_test_date = datetime.datetime(2012, 9, 12, 16, 4, 22)
        end_test_date = datetime.datetime(2012, 9, 13, 23, 59, 59)
        test_filter = logjuggler.date_range_filter(
            start_test_date, end_test_date)
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 4

    def test_date_string_should_be_converted_to_date_object(self, log_lines):
        start_date = '2012-09-12 16:04:22'
        end_date = '2012-09-13 23:59:59'
        test_filter = logjuggler.date_range_filter(
            start_date, end_date)
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 4


class TestLogLevelFilters(object):
    def test_debug_level(self, log_lines):
        test_filter = logjuggler.log_level_filter('DEBUG')
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 3

    def test_info_level(self, log_lines):
        test_filter = logjuggler.log_level_filter('INFO')
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 0

    def test_warn_level(self, log_lines):
        test_filter = logjuggler.log_level_filter('WARN')
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 1

    def test_not_existing_level(self, log_lines):
        test_filter = logjuggler.log_level_filter('xyz')
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 0

    def test_log_level_as_int(self, log_lines):
        test_filter = logjuggler.log_level_filter(234)
        assert len(list(logjuggler.search_results(test_filter, log_lines))) == 0


class TestSessionIdFilter(object):
    def test_existing_session_id(self, log_lines):
        test_session_id = 34523
        test_filter = logjuggler.session_id_filter(sid=test_session_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.session_id for log in search_result]) == 3

    def test_not_existing_session_id(self, log_lines):
        test_session_id = 'We234'
        test_filter = logjuggler.session_id_filter(sid=test_session_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.session_id for log in search_result]) == 0


class TestBusinessIdFilter(object):
    def test_existing_business_id(self, log_lines):
        test_business_id = 1329
        test_filter = logjuggler.business_id_filter(bid=test_business_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.business_id for log in search_result]) == 2

    def test_not_existing_business_id(self, log_lines):
        test_business_id = 'ZxC'
        test_filter = logjuggler.business_id_filter(bid=test_business_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.business_id for log in search_result]) == 0


class TestRequestIdFilter(object):
    def test_existing_request_id(self, log_lines):
        test_request_id = '54ff3'
        test_filter = logjuggler.request_id_filter(rid=test_request_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.request_id for log in search_result]) == 1

    def test_not_existin_grequest_id(self, log_lines):
        test_request_id = '54ff3ssss'
        test_filter = logjuggler.request_id_filter(rid=test_request_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.request_id for log in search_result]) == 0


class TestSearchResults(object):
    def test_log_time_should_be_type_of_str(self, log_lines):
        test_business_id = 1329
        test_filter = logjuggler.business_id_filter(bid=test_business_id)
        search_result = logjuggler.search_results(test_filter, log_lines)
        assert len([log.business_id for log in search_result]) == 2

        search_result = logjuggler.search_results(test_filter, log_lines)
        assert isinstance([item for item in search_result][0].date, str)



logjuggler/logjuggler.py
#!/usr/bin/env python

"""

Logjuggler - a simple log parser.


Usage:
------

    (venvweb)jakub@urababura:~/projects/logjuggler/logjuggler$ python logjuggler.py --file ../data/app.log --help
    usage: logjuggler.py [-h] -f LOGFILE {loglevel,bid,sid,rid,date} ...

    A simple log file parser.

    positional arguments:
        {loglevel,bid,sid,rid,date}
                              Log filters

    optional arguments:
        -h, --help            show this help message and exit
        -f LOGFILE, --file LOGFILE
                              Log file to parse



    (venvweb)jakub@urababura:~/projects/logjuggler/logjuggler$ python logjuggler.py --file ../data/app.log sid 34523
    2012-09-13 16:04:22 DEBUG sid:34523 bid:1329 rid:65d33 message:Starting new session
    2012-09-13 16:04:30 DEBUG sid:34523 bid:1329 rid:54f22 message:Authenticating User
    2012-09-13 16:04:50 ERROR sid:34523 bid:1329 rid:54ff3 message:Missing Authentication token



    (venvweb)jakub@urababura:~/projects/logjuggler/logjuggler$ python logjuggler.py --file ../data/app.log date -h
    usage: logjuggler.py date [-h] start end

    positional arguments:
      start       Start date.
      end         End date.

    optional arguments:
      -h, --help  show this help message and exit



    (venvweb)jakub@urababura:~/projects/logjuggler/logjuggler$ python logjuggler.py --file ../data/app.log date '2012-09-13 16:00:22' '2012-09-13 16:04:45'
    2012-09-13 16:04:22 DEBUG sid:34523 bid:1329 rid:65d33 message:Starting new session
    2012-09-13 16:04:30 DEBUG sid:34523 bid:1329 rid:54f22 message:Authenticating User



    (venvweb)jakub@urababura:~/projects/logjuggler/logjuggler$ python logjuggler.py --file ../data/app.log rid -h
    usage: logjuggler.py rid [-h] rid

    positional arguments:
      rid         Show logs with request id.

    optional arguments:
      -h, --help  show this help message and exit



    (venvweb)jakub@urababura:~/projects/logjuggler/logjuggler$ python logjuggler.py --file ../data/app.log rid 65d33
    2012-09-13 16:04:22 DEBUG sid:34523 bid:1329 rid:65d33 message:Starting new session


"""


import collections
import datetime
import argparse


# namedtuple - storing data from a sinle log line
Log = collections.namedtuple("Log", "date level session_id business_id request_id message")


def read_log_file(file):
    """Returns a log line genarator.

    Args:
        file: str, location of the log file

    Raises:
        IOError if the file can not be found.

    """
    try:
        with open(file, 'r') as f:
            for line in f:
                yield line.strip("\n")
    except IOError:
        print("Log file {file_name} can not be found".format(file_name=file))


def log_message(logline):
    """Return log message (str) from the given logline (str)"""
    return logline[(logline.find("'") + 1): logline.rfind("'")]


def log_level(log_line):
    """Return log level (str) from the given log_line (str)"""
    return log_line.split(" ")[2]


def business_id(log_line):
    """Return business_id (str) from the given log_line (str)"""
    return log_line.split(' ')[4].split(':')[1]


def request_id(log_line):
    """Return request id (str) from the given log_line (str)"""
    return log_line.split(' ')[5].split(':')[1]


def session_id(log_line):
    """Return session id (str) from the given log_line (str)"""
    return log_line.split(' ')[3].split(':')[1]


def log_time(log_line):
    """Return a datetime object from the given log_line (str)"""
    timestamp = ' '.join(log_line.split(" ")[:2])
    return datetime.datetime.strptime(timestamp, '%Y-%m-%d %H:%M:%S')


def time_to_iso(datetime_obj):
    """Return date in iso format (str).

    After parsing log file date is stored as a datetime obj.
    This function allows to change the date back to string
    if necessary (eg, for printing logs, etc)

    """
    return datetime_obj.isoformat(sep=' ')


def time_str_to_datetime(timestring):
    """Return datetime object from the given timestring.

    Args:
        timestamp: str in format: '%Y-%m-%d %H:%M:%S'

    Returns:
        datetime obj

    Raises:
        ValueError, if the string is not well formatted
    """
    try:
        return datetime.datetime.strptime(timestring, '%Y-%m-%d %H:%M:%S')
    except ValueError as e:
        print("Datetime string is malformed. Got exception:\n{0}".format(e))


def search_results(query_filter, logs):
    """Return generator that yields search results

    Args:
        query_filter: func
        logs: generator obj (yields named tuple objects)

    Returns:
        generator obj
    """
    for log in logs:
        result = query_filter(log)
        if result:
            yield convert_to_timestamp(result)


def convert_to_timestamp(tpl):
    """Replace datetime obj with timestamp (str)."""
    return tpl._replace(date=time_to_iso(tpl.date))


def log_level_filter(loglevel):
    """Return a func that filters logs by level.

    Args:
        loglevel: str

    """
    def inner(log_line):
        if log_line.level == str(loglevel).upper():
            return log_line
    return inner


def session_id_filter(sid):
    """Return a func that filters logs by session id.

    Args:
        sid: str

    """
    def inner(log_line):
        if log_line.session_id == str(sid):
            return log_line
    return inner


def business_id_filter(bid):
    """Return func that filters logs by business id.

    Args:
        bid: str
    """
    def inner(log_line):
        if log_line.business_id == str(bid):
            return log_line
    return inner


def request_id_filter(rid):
    """Return func that filters logs by request id.

    Args:
        rid: str

    """
    def inner(log_line):
        if log_line.request_id == str(rid):
            return log_line
    return inner


def date_range_filter(start_date, end_date):
    """Return a func that filter logs between given start and end date.

    Args:
        start_date: timestamp (str) or datetime obj
        end_date: timestamp (str) or datetime obj

    Returns:
        func

    """
    if isinstance(start_date, str):
        start_date = time_str_to_datetime(start_date)
    if isinstance(end_date, str):
        end_date = time_str_to_datetime(end_date)

    def inner(log_line):
        if start_date <= log_line.date <= end_date:
            return log_line
    return inner


def display_log(log):
    """Print a log line based on defined template.

    Args:
        log: namedtuple log obj

    """
    template = ("{date} {level} sid:{session_id} bid:{business_id} "
                "rid:{request_id} message:{message}")
    print template.format(date=log.date, level=log.level, session_id=log.session_id,
                          business_id=log.business_id, request_id=log.request_id,
                          message=log.message)


def display_search_results(results):
    """Print log lines from results.

    Args:
        results: list of filtered logs
    """
    for log in results:
        display_log(log)


# functions for profiling with decorator

def get_log_level(log_level, log_entries):
    """Return a list with log leve search results."""
    return [res for res in search_results(log_level_filter(log_level), log_entries)]


def get_sid(sid, log_entries):
    """Return a list with sessionid search results."""
    return [res for res in search_results(session_id_filter(sid), log_entries)]


def get_bid(bid, log_entries):
    """Return a list wits business id search results."""
    return [res for res in search_results(business_id_filter(bid), log_entries)]


def get_rid(rid, log_entries):
    """Return a list with request id rearch results."""
    return [res for res in search_results(request_id_filter(rid), log_entries)]


def get_dates(start_date, end_date, log_entries):
    """Return a list with date range seach results."""
    return [res for res in (search_results(date_range_filter(
        start_date=arg_dict.get('start'), end_date=arg_dict.get('end')), log_entries))
    ]


if __name__ == "__main__":

    parser = argparse.ArgumentParser(description="A simple log file parser.")
    parser.add_argument('-f', '--file', dest='logfile', action='store',
                        help='Log file to parse', required=True)

    subparsers = parser.add_subparsers(help='Log filters')

    log_level_parser = subparsers.add_parser('loglevel')
    log_level_parser.add_argument('loglevel', action='store',
                                  choices=('DEBUG', 'INFO', 'WARN', 'ERROR'),
                                  help="Show logs with given loglevel.")

    business_id_parser = subparsers.add_parser('bid')
    business_id_parser.add_argument('bid', action='store',
                                    help='Show logs with given business id')

    session_id_parser = subparsers.add_parser('sid')
    session_id_parser.add_argument('sid', action='store',
                                   help='Show logs with session id')

    request_id_parser = subparsers.add_parser('rid')
    request_id_parser.add_argument('rid', action='store',
                                   help='Show logs with request id.')

    date_parser = subparsers.add_parser('date')
    date_parser.add_argument('start', action='store', help='Start date.')
    date_parser.add_argument('end', action='store', help='End date.')

    arg_dict = vars(parser.parse_args())

    log_entries = (Log(log_time(l), log_level(l), session_id(l), business_id(l),
                       request_id(l), log_message(l)) for l in read_log_file(arg_dict.get('logfile')))

    if arg_dict.get('loglevel'):
        result = get_log_level(arg_dict.get('loglevel'), log_entries)
        display_search_results(result)

    if arg_dict.get('bid'):
        result = get_bid(arg_dict.get('bid'), log_entries)
        display_search_results(result)

    if arg_dict.get('sid'):
        result = get_sid(arg_dict.get('sid'), log_entries)
        display_search_results(result)

    if arg_dict.get('rid'):
        result = get_rid(arg_dict.get('rid'), log_entries)
        display_search_results(result)

    if arg_dict.get('start') and arg_dict.get('end'):
        result = get_dates(arg_dict.get('start'), arg_dict.get('end'), log_entries)
        display_search_results(result)


logjuggler/logprofiler.py
#!/usr/bin/env python

"""

Example of a profiler decorator that collect basic
stats about decorated function.

Usage:
    >>>
    >>> from logprofiler import profile
    >>>
    >>> @profile
    ... def sample_foo():
        ... return [item for item in xrange(10)]
        ...
    >>> sample_foo()

        === Function profiler report ===

        Function:   sample_foo
        NumSamples: 1
        Min:        2.47955322266e-05
        Max         2.47955322266e-05
        Average:    2.47955322266e-05

        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

    >>> sample_foo()

        === Function profiler report ===

        Function:   sample_foo
        NumSamples: 2
        Min:        1.78813934326e-05
        Max         2.47955322266e-05
        Average:    2.13384628296e-05

        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    >>>
    >>> sample_foo()

        === Function profiler report ===

        Function:   sample_foo
        NumSamples: 3
        Min:        1.78813934326e-05
        Max         2.47955322266e-05
        Average:    2.121925354e-05

        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    >>>
    >>> sample_foo()

        === Function profiler report ===

        Function:   sample_foo
        NumSamples: 4
        Min:        1.78813934326e-05
        Max         2.47955322266e-05
        Average:    2.03847885132e-05

        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    >>>


Usage:
    >>> import logjuggler as lj
    >>> from logprofiler import profile
    >>>
    >>> log_entries = [lj.Log(lj.log_time(l), lj.log_level(l), lj.session_id(l), lj.business_id(l), lj.request_id(l), lj.log_message(l)) for l in lj.read_log_file('../data/app.log')]
    >>> log_entries
    [Log(date=datetime.datetime(2012, 9, 13, 16, 4, 22), level='DEBUG', session_id='34523', business_id='1329', request_id='65d33', message='Starting new session'), Log(date=datetime.datetime(2012, 9, 13, 16, 4, 30), level='DEBUG', session_id='34523', business_id='1329', request_id='54f22', message='Authenticating User'), Log(date=datetime.datetime(2012, 9, 13, 16, 5, 30), level='DEBUG', session_id='42111', business_id='319', request_id='65a23', message='Starting new session'), Log(date=datetime.datetime(2012, 9, 13, 16, 4, 50), level='ERROR', session_id='34523', business_id='1329', request_id='54ff3', message='Missing Authentication token'), Log(date=datetime.datetime(2012, 9, 13, 16, 5, 31), level='DEBUG', session_id='42111', business_id='319', request_id='86472', message='Authenticating User'), Log(date=datetime.datetime(2012, 9, 13, 16, 5, 31), level='DEBUG', session_id='42111', business_id='319', request_id='7a323', message='Deleting asset with ID 543234'), Log(date=datetime.datetime(2012, 9, 13, 16, 5, 32), level='WARN', session_id='42111', business_id='319', request_id='7a323', message='Invalid asset ID')]
    >>>
    >>>
    >>> @profile
    ... def log_level(log_level, log_entries):
    ...     lj.get_log_level(log_level, log_entries)
    >>>
    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   1
    Min:      0.000174045562744
    Max       0.000174045562744
    Average:  0.000174045562744

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   2
    Min:      0.000169038772583
    Max       0.000174045562744
    Average:  0.000171542167664

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   3
    Min:      0.000169038772583
    Max       0.000227928161621
    Average:  0.000190337498983

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   4
    Min:      0.000169038772583
    Max       0.000227928161621
    Average:  0.000186502933502

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   5
    Min:      0.000169038772583
    Max       0.000227928161621
    Average:  0.000184011459351

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   6
    Min:      0.000169038772583
    Max       0.000227928161621
    Average:  0.000181992848714

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   7
    Min:      0.000169038772583
    Max       0.000235080718994
    Average:  0.000189576830183

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   8
    Min:      0.000169038772583
    Max       0.000235080718994
    Average:  0.000187516212463

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   9
    Min:      0.000168085098267
    Max       0.000235080718994
    Average:  0.000185357199775

    >>> log_level('DEBUG', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   10
    Min:      0.000168085098267
    Max       0.000235080718994
    Average:  0.000184416770935

    >>> log_level('WARN', log_entries)

    === Function profiler report ===

    Function: log_level
    NumSamples:   11
    Min:      0.000102043151855
    Max       0.000235080718994
    Average:  0.00017692826011


"""

import time


class profile(object):

    """Profiler decorator."""

    def __init__(self, func):
        self.func = func
        self.counter = 0
        self.executions = []
        self.stats = {'func': self.func.__name__}

    def __call__(self, *args, **kwargs):
        start = time.time()

        result = self.func(*args, **kwargs)

        elapsed_time = time.time() - start
        self.counter += 1
        self.add_exec_sample(elapsed_time)
        self.update_stats()
        self.print_report()
        return result

    def add_exec_sample(self, elapsed):
        self.executions.append(elapsed)

    def update_stats(self):
        self.stats['counter'] = self.counter
        self.stats['avg_time'] = sum(self.executions) / self.counter
        self.stats['min_time'] = min(self.executions)
        self.stats['max_time'] = max(self.executions)

    def print_report(self):
        template = ("\n=== Function profiler report ===\n\n"
                    "Function:\t{func}\nNumSamples:\t{counter}\n"
                    "Min:\t\t{tmin}\nMax\t\t{tmax}\nAverage:\t{tavg}\n")
        print template.format(func=self.stats.get('func'),
                              counter=self.stats.get('counter'),
                              tmin=self.stats.get('min_time'),
                              tmax=self.stats.get('max_time'),
                              tavg=self.stats.get('avg_time'))


@profile
def test_function():
    return [a for a in xrange(10)]



if __name__=="__main__":
    test_function()





